### DISCUSSION TITLE
Data values in the initial state are rounded

### DISCUSSION THREAD

**Discussion by BMP-TUD** [zwicker-group/py-pde#233] [General]
Dear Dr Zwicker, 

Thank you for the great package, but I have a question concerning the data formats: When I load an initial state with the `Scalar_Field.from_image` function and I call `.data` than the results are roughly correct, but everything that is behind a comma is rounded. Is this intentional or a bug? Thank you for your help in advance, 

Best wishes, 
Bartosz

**Comment by david-zwicker** [comment]
I think this might simply be an issue with the display style. Does calling `numpy.set_printoptions(precision=15)` before you print the `data` attribute help?

**Reply by BMP-TUD** [reply]
Thanks for the quick response, it actually does not help, output is the following: 

```
np.set_printoptions(precision=15)

state.data
array([[ 1.,  1.,  1., ...,  1.,  1.,  1.],
       [ 1.,  1.,  1., ...,  1.,  1.,  1.],
       [ 1.,  1.,  1., ...,  1.,  1.,  1.],
       ...,
       [ 1.,  1.,  1., ..., 11., 10.,  9.],
       [ 1.,  1.,  1., ...,  9.,  9.,  9.],
       [ 1.,  1.,  1., ...,  8.,  8.,  8.]])
```

but it should be: 
```
u
array([[ 1.415278196334839,  1.410358190536499,  1.426051259040833, ...,
         9.334187507629395,  9.264062881469727,  8.686315536499023],
       [ 1.648768186569214,  1.66365647315979 ,  1.695865750312805, ...,
        10.378127098083496,  9.108780860900879,  8.364289283752441],
       [ 1.62944757938385 ,  1.662785649299622,  1.704971432685852, ...,
        11.031474113464355,  9.501060485839844,  8.784432411193848],
       ...,
       [ 1.27515983581543 ,  1.260051012039185,  1.252041459083557, ...,
         1.130288124084473,  1.174511432647705,  1.178279638290405],
       [ 1.275601506233215,  1.277632474899292,  1.279035210609436, ...,
         1.167059898376465,  1.179918646812439,  1.176928877830505],
       [ 1.271483778953552,  1.27246105670929 ,  1.279974102973938, ...,
         1.171810865402222,  1.184486389160156,  1.18511164188385 ]])
```

Could there be a problem if I load the data from a `.tif`?

**Reply by david-zwicker** [reply]
Interesting. The function to read images loads the file using `matplotlib.pyplot.imread`. Could you test whether this function works on your image?

**Reply by BMP-TUD** [reply]
This is probably causing the problem, as the results are rounded and the image is loaded as a `uint8`  and gives this:
```
test

array([[[  1,   1,   1, 255],
        [  1,   1,   1, 255],
        [  1,   1,   1, 255],
        ...,
        [  9,   9,   9, 255],
        [  9,   9,   9, 255],
        [  8,   8,   8, 255]],
        ...,
        [  1,   1,   1, 255],
        [  1,   1,   1, 255],
        [  1,   1,   1, 255]]], dtype=uint8)
```
But my original image is of type `float32` if read in with `tifffile.imread`. I checked that pyplot is using either `.png` format or the format is detected by `PIL.Image.open`. If I run this than the format is still correct: `<PIL.TiffImagePlugin.TiffImageFile image mode=F size=186x138 at 0x1E3296EA0C8>`  but `matplotlib.pyplot.imread` is not recognizing it correctly. I changed the `imread` function in `scalar.py` to that from `tifffile` and the data is read in correctly. Maybe there must be an additional check for the format, if it is a `.tif` and you could implement `tifffile.imread` instead? 

Anyway thanks for your help :)

**Reply by david-zwicker** [reply]
This sounds like a good idea! Could you create a branch/fork with both options for reading images (e.g., using an argument to distinguish different variants or detecting the right reader based on file extension)? I'm happy to accept pull requests and integrate this into the main branch, but I won't have time to work on it myself during the next weeks.
